pkg/apiserver/resthandler.go

```diff
@@ -171,6 +172,85 @@ func CreateResource(r RESTCreater, ctxFn ContextFunc, namer ScopeNamer, codec ru
 	}
 }
 
+// PatchResource returns a function that will handle a resource patch
+func PatchResource(r RESTPatcher, ctxFn ContextFunc, namer ScopeNamer, codec runtime.Codec, resource string, admit admission.Interface) restful.RouteFunction {
+	return func(req *restful.Request, res *restful.Response) {
+		w := res.ResponseWriter
+
+		// TODO: we either want to remove timeout or document it (if we document, move timeout out of this function and declare it in api_installer)
+		timeout := parseTimeout(req.Request.URL.Query().Get("timeout"))
+
+		namespace, name, err := namer.Name(req)
+		if err != nil {
+			notFound(w, req.Request)
+			return
+		}
+		ctx := ctxFn(req)
+		ctx = api.WithNamespace(ctx, namespace)
+
+		original, err := r.Get(ctx, name)
```

Comment:
This results in two successive reads that are not controlled by the resource version, which I think is wrong.  The client has to be able to specify which version to check against.  So there's at least a check missing here for the input version against the get version.

I'm mentally debating whether passing a structured type down into Update would be more appropriate (delta, with attached api version).  Even a generic patch resource being passed down would work.  I mildly dislike reading here then reading again on update - I'd prefer to do it exactly once, in the storage.  That's something we could fix later, but it still seems messy.